lightningd/peer_htlcs.c

```diff
@@ -193,6 +242,34 @@ void local_fail_in_htlc(struct htlc_in *hin, const u8 *failmsg TAKES)
 	fail_in_htlc(hin, take(failonion));
 }
 
+/* This is used for cases where we can immediately fail the HTLC, but
+ * need to append a channel_update. */
+void local_fail_in_htlc_needs_update(struct htlc_in *hin,
+				     const u8 *failmsg_needs_update TAKES,
+				     const struct short_channel_id *failmsg_scid)
+{
+	struct failmsg_update_cbdata *cbdata;
+
+	/* To avoid the state where we have no failonion, we use a temporary
+	 * one, and update once we get the reply from gossipd. */
+	assert(!hin->preimage);
+
+	hin->failonion = create_onionreply(hin,
+					   hin->shared_secret,
+					   towire_temporary_node_failure(tmpctx));
+	/* We update state now to signal it's in progress, for persistence. */
+	htlc_in_update_state(hin->key.channel, hin, SENT_REMOVE_HTLC);
+	htlc_in_check(hin, __func__);
+
+	cbdata = tal(hin, struct failmsg_update_cbdata);
+	cbdata->hin = hin;
+	cbdata->failmsg_needs_update
+		= tal_dup_talarr(cbdata, u8, failmsg_needs_update);
+	subd_req(cbdata, hin->key.channel->peer->ld->gossip,
+		 take(towire_gossip_get_stripped_cupdate(NULL, failmsg_scid)),
+		 -1, 0, failmsg_update_reply, cbdata);
```

Comment:
seems like your `gossip` daemon doesnt exit? failing `test_restart_many_payments`

```
EBUG:root:lightningd-1: 2020-02-25T06:08:23.095Z DEBUG 0382ce59ebf18be7d84677c2e35f23294b9992ceca95491fcf8a56c6cb2d9de199-chan#3: HTLC in 1 RCVD_ADD_ACK_REVOCATION->SENT_REMOVE_HTLC
DEBUG:root:lightningd-1: 2020-02-25T06:08:23.108Z **BROKEN** lightningd: FATAL SIGNAL 11 (version v0.8.1-36-g03045e1-modded)
DEBUG:root:lightningd-1: 2020-02-25T06:08:23.108Z **BROKEN** lightningd: backtrace: common/daemon.c:44 (send_backtrace) 0x5634d3e4c1b0
DEBUG:root:lightningd-1: 2020-02-25T06:08:23.108Z **BROKEN** lightningd: backtrace: common/daemon.c:52 (crashdump) 0x5634d3e4c200
DEBUG:root:lightningd-1: 2020-02-25T06:08:23.108Z **BROKEN** lightningd: backtrace: (null):0 ((null)) 0x7fb9d3ca4f1f
DEBUG:root:lightningd-1: 2020-02-25T06:08:23.108Z **BROKEN** lightningd: backtrace: lightningd/subd.c:750 (subd_send_msg) 0x5634d3e43132
DEBUG:root:lightningd-1: 2020-02-25T06:08:23.108Z **BROKEN** lightningd: backtrace: lightningd/subd.c:770 (subd_req_) 0x5634d3e431f6
DEBUG:root:lightningd-1: 2020-02-25T06:08:23.108Z **BROKEN** lightningd: backtrace: lightningd/peer_htlcs.c:268 (local_fail_in_htlc_needs_update) 0x5634d3e35438
DEBUG:root:lightningd-1: 2020-02-25T06:08:23.108Z **BROKEN** lightningd: backtrace: lightningd/peer_htlcs.c:296 (fail_out_htlc) 0x5634d3e3558d
DEBUG:root:lightningd-1: 2020-02-25T06:08:23.108Z **BROKEN** lightningd: backtrace: lightningd/peer_htlcs.c:516 (destroy_hout_subd_died) 0x5634d3e35c86
DEBUG:root:lightningd-1: 2020-02-25T06:08:23.108Z **BROKEN** lightningd: backtrace: ccan/ccan/tal/tal.c:240 (notify) 0x5634d3eb2472
DEBUG:root:lightningd-1: 2020-02-25T06:08:23.108Z **BROKEN** lightningd: backtrace: ccan/ccan/tal/tal.c:402 (del_tree) 0x5634d3eb2961
DEBUG:root:lightningd-1: 2020-02-25T06:08:23.108Z **BROKEN** lightningd: backtrace: ccan/ccan/tal/tal.c:486 (tal_free) 0x5634d3eb2ced
DEBUG:root:lightningd-1: 2020-02-25T06:08:23.108Z **BROKEN** lightningd: backtrace: lightningd/peer_htlcs.c:2107 (free_htlcs) 0x5634d3e3a85a
DEBUG:root:lightningd-1: 2020-02-25T06:08:23.108Z **BROKEN** lightningd: backtrace: lightningd/lightningd.c:520 (shutdown_subdaemons) 0x5634d3e1a7a1
DEBUG:root:lightningd-1: 2020-02-25T06:08:23.108Z **BROKEN** lightningd: backtrace: lightningd/lightningd.c:940 (main) 0x5634d3e1b217
DEBUG:root:lightningd-1: 2020-02-25T06:08:23.108Z **BROKEN** lightningd: backtrace: (null):0 ((null)) 0x7fb9d3c87b96
DEBUG:root:lightningd-1: 2020-02-25T06:08:23.108Z **BROKEN** lightningd: backtrace: (null):0 ((null)) 0x5634d3e01b09
DEBUG:root:lightningd-1: 2020-02-25T06:08:23.108Z **BROKEN** lightningd: backtrace: (null):0 ((null)) 0xffffffffffffffff
DEBUG:root:Received response for stop call: {'error': 'Connection to RPC server lost.'}
----------------------------- Captured stderr call -----------------------------
lightningd: FATAL SIGNAL 11 (version v0.8.1-36-g03045e1-modded)
0x5634d3e4c15a send_backtrace
	common/daemon.c:39
0x5634d3e4c200 crashdump
	common/daemon.c:52
0x7fb9d3ca4f1f ???
	???:0
0x5634d3e43132 subd_send_msg
	lightningd/subd.c:750
0x5634d3e431f6 subd_req_
	lightningd/subd.c:770
0x5634d3e35438 local_fail_in_htlc_needs_update
	lightningd/peer_htlcs.c:268
0x5634d3e3558d fail_out_htlc
	lightningd/peer_htlcs.c:296
0x5634d3e35c86 destroy_hout_subd_died
	lightningd/peer_htlcs.c:516
0x5634d3eb2472 notify
	ccan/ccan/tal/tal.c:240
0x5634d3eb2961 del_tree
	ccan/ccan/tal/tal.c:402
0x5634d3eb2ced tal_free
	ccan/ccan/tal/tal.c:486
0x5634d3e3a85a free_htlcs
	lightningd/peer_htlcs.c:2107
0x5634d3e1a7a1 shutdown_subdaemons
	lightningd/lightningd.c:520
0x5634d3e1b217 main
	lightningd/lightningd.c:940
0x7fb9d3c87b96 ???
	???:0
0x5634d3e01b09 ???
	???:0
0xffffffffffffffff ???
	???:0
```

Reply:
Ah, I think this is best addressed by freeing HTLCs and peer daemons first, then gossipd.  Thanks!